### DISCUSSION TITLE
Pressing ctrl+c should close the command prompt

### DISCUSSION THREAD

**Discussion by pinpox** [neomutt/neomutt#4497] [Development]
Pressing `ctrl+c` should close the command prompt, as it does in vim.

@flatcap over on IRC:
`Ctrl-C is trapped by the terminal and turned into SIGINT, which neomutt traps and sets the variable SigInt  it _might_ be as simple as checking SigInt (and resetting it)
`

**Comment by steef435** [comment]
I did some digging.

mutt_enter_command (commands.c) calls mutt_get_field (curs_lib.c), which calls mutt_enter_string_full (enter.c), which calls km_dokey (keymap.c), which finally calls mutt_getch (curs_lib.c) and this is the function that calls mutt_query_exit on SigInt (and currently no way to configure this behavior).

The only "nice" way I can think of to implement this is to make it configurable by function argument. Together, the above functions are used more than 50 times in the code base, so for all those calls the desired behavior would have to be evaluated.

So yeah, it's as simple as checking SigInt, but it's buried so deeply in the code hierarchy that I think it will take some work (I'm not familiar enough to judge which behavior is desired for each call).

**Comment by flatcap** [comment]
Nice investigating!

Before you start altering code, I'd just like to get a second opinion from the other developers.
We already have <kbd>Ctrl-G</kbd> (a scarcely documented abort key).

---

Yes, a parameter on `mutt_getch()` would work to distinguish the interruptible case (from normal keypresses).
There _are_ quite a lot of calls to `mutt_getch()`, but we only want to interrupt the full text-entry and it seems that they all come from one place `mutt_get_field_full()`:

![get-field](https://flatcap.org/mutt/get-field.svg)

**Comment by pinpox** [comment]
Sorry, have been gone for a while. I'm still interested in this and would be willing to help.
Does that mean that only `mutt_get_field_full()` would have to be modified?

EDIT: Also, how is this related to  #1510, can both issues be solved in one go?

**Comment by flatcap** [comment]
> Sorry, have been gone for a while

No worries.

> Does that mean that only `mutt_get_field_full()` would have to be modified?

I _think_ so.

> how is this related to #1510

It's the same sort of thing.

@neomutt/reviewers 
For text-entry fields, what do we want to (continue) support(ing)?
- <kbd>Ctrl-G</kbd>
- <kbd>Ctrl-C</kbd>
- <kbd>Escape</kbd>

**Comment by gahr** [comment]
I'd love to have them all

**Comment by steef435** [comment]
I tried to solve #1510 by binding Esc to a new OP_EDITOR_CANCEL opcode (which get_field_full then interprets), which works, but since Esc has the same value as Alt, this then shades all the other EDITOR bindings using Alt as a modifier. (like Alt-u for capitalizing a word). It doesn't look like the current implementation of km_dokey and mutt_getch differentiate between a stand-alone press or a combination. (and so far I haven't found a nice way to do so with curses either)

EDIT: what I'm trying to say is, maybe there are better ways to handle #1510 without hard-coding yet another key in mutt_getch, since we're not dealing with SIGINT there)

**Comment by flatcap** [comment]
Ah!  That's unfortunate, but not unexpected.
I haven't really read/understood the key-handling code in (Neo)Mutt -- no doubt it could do with an overhaul.

To separate <kbd>Escape</kbd> from <kbd>Escape</kbd>-<kbd>t</kbd>, we'd need to add timing info to the keyboard events.  Vim has `ttimeout` and `ttimeoutlen` (50ms).  If the second character arrives within that timeframe, then they're one entity.

Then, we could match those entities to NeoMutt bindings, etc.

**Comment by ftilde** [comment]
If #2139 is accepted, imho the best solution here would be to call `raw()` instead of `cbreak()` during TUI initialization. Then, curses does not raise SIGINT on ctrl-c anymore and we can simply use it as a key binding. (I have just tested this locally.)

Quoting from the [man page](https://linux.die.net/man/3/cbreak):

> The differences are that in raw mode, the interrupt, quit, suspend, and flow control characters are all passed through uninterpreted, instead of generating a signal. 

As a result, we would have to add actions like `<ask-quit>` (ctrl-c), `<suspend>` (ctrl-z) that should be bound to the corresponding keys in the default config, but can of course be overwritten.

**Comment by c02y** [comment]
When I start neomutt and for some reason it won't load my emails list such as stuck with 
~~~
gpg: AES encrypted data
gpg: encrypted with 1 passphrase
~~~
or asking for password, and I want to quit neomutt with Ctrl-c, it just stucks in there and I have to kill it in another terminal manually.
